src/watcherEx.ts

```diff
@@ -0,0 +1,227 @@
+// Copyright 2022 The Casbin Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { RedisClusterConnection, RedisConnection, RedisClient } from './redis';
+import { ClusterNode, ClusterOptions, RedisOptions } from 'ioredis';
+import { Model, WatcherEx } from 'casbin';
+
+export interface WatcherExOptions extends RedisOptions {
+  channel?: string;
+}
+
+export interface WatcherExClusterOptions extends ClusterOptions {
+  channel?: string;
+}
+
+export interface WatcherExCallBackOptions {
+  addPolicyCb?: any;
+  removePolicyCb?: any;
+  removeFilteredPolicyCb?: any;
+  savePolicyCb?: any;
+  addPoliciesCb?: any;
+  removePoliciesCb?: any;
+}
+
+const updateTypes = {
+  addPolicy: 'updateForAddPolicy',
+  removePolicy: 'updateForRemovePolicy',
+  removeFilteredPolicy: 'updateForRemoveFilteredPolicy',
+  savePolicy: 'updateForSavePolicy',
+  addPolicies: 'updateForAddPolicies',
+  removePolicies: 'updateForRemovePolicies'
+};
+
+export class RedisWatcherEx implements WatcherEx {
+  private pubConnection: RedisConnection | RedisClusterConnection;
+  private subConnection: RedisConnection | RedisClusterConnection;
+  private watcherExCallBackOptions?: WatcherExCallBackOptions;
+  private channel = 'casbin';
+
+  /**
+   * newWatcherEx creates a watcherEx on the single Redis.
+   * @param options - https://github.com/luin/ioredis/blob/master/API.md#new-redisport-host-options
+   * @example
+   * newWatcherEx('redis://user:password@redis-service.com:6379')
+   * newWatcherEx('//localhost:6379')
+   * newWatcherEx({
+   *   port: 6379, // Redis port
+   *   host: "127.0.0.1", // Redis host
+   *   family: 4, // 4 (IPv4) or 6 (IPv6)
+   *   password: "auth",
+   *   db: 0,
+   *   channel: "casbin"
+   * })
+   */
+  public static async newWatcherEx(
```

Comment:
I guess initiators for this and Cluster could be combined. Not a big deal.

Reply:
It was separated in watcher, so I keep the same here, but then should I also combine them in watcher?

Reply:
Golang definition seems to have them separate, so let's keep them separate in both.